File: smali/src/main/java/com/github/netomi/bat/smali/LocalVariableCollector.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.smali;

import com.github.netomi.bat.dexfile.DexFile;
import com.github.netomi.bat.dexfile.debug.*;
import com.github.netomi.bat.dexfile.visitor.DebugSequenceVisitor;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static com.github.netomi.bat.dexfile.DexConstants.NO_INDEX;

class      LocalVariableCollector
implements DebugSequenceVisitor
{
    private final Map<Integer, List<String>> debugState;
    private final RegisterPrinter            registerPrinter;
    private final LocalVariableInfo[]        localVariableInfos;

    private short codeOffset;

    LocalVariableCollector(Map<Integer, List<String>> debugState, LocalVariableInfo[] localVariableInfos, RegisterPrinter registerPrinter) {
        this.debugState         = debugState;
        this.registerPrinter    = registerPrinter;
        this.localVariableInfos = localVariableInfos;
    }

    @Override
    public void visitAnyDebugInstruction(DexFile dexFile, DebugInfo debugInfo, DebugInstruction instruction) {}

    @Override
    public void visitAdvanceLineAndPC(DexFile dexFile, DebugInfo debugInfo, DebugAdvanceLineAndPC instruction) {
        codeOffset += instruction.getAddrDiff();
    }

    @Override
    public void visitAdvancePC(DexFile dexFile, DebugInfo debugInfo, DebugAdvancePC instruction) {
        codeOffset += instruction.getAddrDiff();
    }

    @Override
    public void visitEndLocal(DexFile dexFile, DebugInfo debugInfo, DebugEndLocal instruction) {
        StringBuilder sb = new StringBuilder();

        int registerNum = instruction.getRegisterNum();
        sb.append(".end local ");
        sb.append(registerPrinter.formatRegister(registerNum));

        handleGenericLocal(registerNum, sb);

        addDebugInfo(codeOffset, sb.toString());
    }

    @Override
    public void visitRestartLocal(DexFile dexFile, DebugInfo debugInfo, DebugRestartLocal instruction) {
        StringBuilder sb = new StringBuilder();

        int registerNum = instruction.getRegisterNum();
        sb.append(".restart local ");
        sb.append(registerPrinter.formatRegister(registerNum));

        handleGenericLocal(registerNum, sb);

        addDebugInfo(codeOffset, sb.toString());
    }

    @Override
    public void visitStartLocal(DexFile dexFile, DebugInfo debugInfo, DebugStartLocal instruction) {
        handleStartLocal(dexFile,
                         instruction.getRegisterNum(),
                         instruction.getNameIndex(),
                         instruction.getTypeIndex(),
                         NO_INDEX);
    }

    @Override
    public void visitStartLocalExtended(DexFile dexFile, DebugInfo debugInfo, DebugStartLocalExtended instruction) {
        handleStartLocal(dexFile,
                         instruction.getRegisterNum(),
                         instruction.getNameIndex(),
                         instruction.getTypeIndex(),
                         instruction.getSigIndex());
    }

    private void handleStartLocal(DexFile dexFile, int registerNum, int nameIndex, int typeIndex, int sigIndex) {
        localVariableInfos[registerNum] =
                new LocalVariableInfo(dexFile.getString(nameIndex),
                                      dexFile.getType(typeIndex),
                                      dexFile.getString(sigIndex));

        StringBuilder sb = new StringBuilder();

        sb.append(".local ");
        sb.append(registerPrinter.formatRegister(registerNum));
        sb.append(", ");

        String name = dexFile.getString(nameIndex);
        if (name != null) {
            sb.append("\"");
            sb.append(name);
            sb.append("\"");
        } else {
            sb.append("null");
        }
        sb.append(":");
        sb.append(dexFile.getType(typeIndex));

        if (sigIndex != NO_INDEX) {
            sb.append(", \"");
            sb.append(dexFile.getString(sigIndex));
            sb.append("\"");
        }

        addDebugInfo(codeOffset, sb.toString());
    }

    private void handleGenericLocal(int registerNum, StringBuilder sb) {
        LocalVariableInfo localVariableInfo = localVariableInfos[registerNum];
        if (localVariableInfo != null) {
            sb.append("    # ");

            if (localVariableInfo.name != null) {
                sb.append("\"");
                sb.append(localVariableInfo.name);
                sb.append("\"");
            } else {
                sb.append("null");
            }

            sb.append(":");
            sb.append(localVariableInfo.type);

            if (localVariableInfo.signature != null) {
                sb.append(", \"");
                sb.append(localVariableInfo.signature);
                sb.append("\"");
            }
        }
    }

    private void addDebugInfo(int offset, String info) {
        List<String> infos = debugState.computeIfAbsent(offset, (key) -> new ArrayList<>());
        infos.add(info);
    }

    public static class LocalVariableInfo
    {
        final String name;
        final String type;
        final String signature;

        LocalVariableInfo(String name, String type, String signature) {
            this.name      = name;
            this.type      = type;
            this.signature = signature;
        }
    }
}

File: smali/src/main/kotlin/com/github/netomi/bat/smali/disassemble/LocalVariableCollector.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.smali.disassemble

import com.github.netomi.bat.dexfile.DexFile
import com.github.netomi.bat.dexfile.NO_INDEX
import com.github.netomi.bat.dexfile.debug.*
import com.github.netomi.bat.dexfile.debug.visitor.DebugSequenceVisitor

internal class LocalVariableCollector(private val debugState:         MutableMap<Int, MutableList<String>>,
                                      private val localVariableInfos: Array<LocalVariableInfo?>,
                                      private val registerPrinter:    RegisterPrinter) : DebugSequenceVisitor {

    private var codeOffset: Int = 0

    override fun visitAnyDebugInstruction(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugInstruction) {}

    override fun visitAdvanceLineAndPC(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugAdvanceLineAndPC) {
        codeOffset += instruction.addrDiff
    }

    override fun visitAdvancePC(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugAdvancePC) {
        codeOffset += instruction.addrDiff
    }

    override fun visitEndLocal(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugEndLocal) {
        val info = buildString {
            val registerNum = instruction.registerNum
            append(".end local ")
            append(registerPrinter.formatRegister(registerNum))
            handleGenericLocal(registerNum, this)
        }
        addDebugInfo(codeOffset, info)
    }

    override fun visitRestartLocal(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugRestartLocal) {
        val info = buildString {
            val registerNum = instruction.registerNum
            append(".restart local ")
            append(registerPrinter.formatRegister(registerNum))
            handleGenericLocal(registerNum, this)
        }
        addDebugInfo(codeOffset, info)
    }

    override fun visitStartLocal(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugStartLocal) {
        handleStartLocal(dexFile, instruction.registerNum, instruction.nameIndex, instruction.typeIndex, NO_INDEX)
    }

    override fun visitStartLocalExtended(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugStartLocalExtended) {
        handleStartLocal(dexFile, instruction.registerNum, instruction.nameIndex, instruction.typeIndex, instruction.sigIndex)
    }

    private fun handleStartLocal(dexFile: DexFile, registerNum: Int, nameIndex: Int, typeIndex: Int, sigIndex: Int) {
        localVariableInfos[registerNum] = LocalVariableInfo(
            dexFile.getStringNullable(nameIndex),
            dexFile.getTypeOrNull(typeIndex)?.type,
            dexFile.getStringNullable(sigIndex)
        )

        val info = buildString {
            append(".local ")
            append(registerPrinter.formatRegister(registerNum))
            append(", ")

            append("\"")
            if (nameIndex != NO_INDEX) {
                append(dexFile.getString(nameIndex))
            }
            append("\"")

            if (typeIndex != NO_INDEX) {
                append(":")
                append(dexFile.getType(typeIndex))
            }

            if (sigIndex != NO_INDEX) {
                append(", \"")
                append(dexFile.getString(sigIndex))
                append("\"")
            }
        }
        addDebugInfo(codeOffset, info)
    }

    private fun handleGenericLocal(registerNum: Int, sb: StringBuilder) {
        val localVariableInfo = localVariableInfos[registerNum]
        localVariableInfo?.apply {
            sb.append("    # ")
            if (name != null) {
                sb.append("\"")
                sb.append(name)
                sb.append("\"")
            } else {
                sb.append("null")
            }
            sb.append(":")
            sb.append(type)
            if (signature != null) {
                sb.append(", \"")
                sb.append(signature)
                sb.append("\"")
            }
        }
    }

    private fun addDebugInfo(offset: Int, info: String) {
        val infos = debugState.computeIfAbsent(offset) { ArrayList() }
        infos.add(info)
    }
}

internal class LocalVariableInfo internal constructor(val name: String?, val type: String?, val signature: String?)
